### PR TITLE
fix(verifier,codegen): guard @Nat->@Int widening at the per-component target-type sites (#820)

### LINKED ISSUE #820: @Nat → @Int widening: remaining deferred coercion sites (follow-up to #813)
## Summary

The `@Nat → @Int` widening soundness work ([#813](https://github.com/aallan/vera/issues/813)) guards every site where code generation can statically determine the source is `@Nat` — return, `let`, call argument, concrete `@Int` constructor field, ADT sub-pattern extraction, match binding — and its follow-up in **#821** added two more: the explicit `nat_to_int` built-in and heterogeneous `if`/`match` arms whose alternative is a **non-negative literal**. This issue tracks the sites that remain deferred.

### Already fixed (do not re-do)

- **`nat_to_int(@Nat)` built-in** (was section D below) — its declared `@Int` return had masked the `@Nat` source, so `vera verify` falsely proved `ensures(@Int.result >= 0)` while `vera run -- 18446744073709551615` returned `-1`. Fixed in #821; it now carries the `nat_to_int_coerce` obligation + the boundary trap guard, which caught real unguarded widenings in `json.vera` / `life.vera`.
- **Heterogeneous `if`/`match` with a non-negative-literal alternative** (`if c then { @Nat.0 } else { 0 }`) — part of section C below. Fixed in #821 by treating a non-negative literal arm as `@Nat`-compatible, so the whole expression stays `@Nat` and the single boundary guard covers the real `@Nat` arm.

### Moved to the narrowing-dual issues

Three deferred widening sites share the **exact** blocker of their `@Int → @Nat` narrowing duals (code generation has no per-component target-type metadata), so they are tracked alongside the narrowing direction rather than duplicated here:

- **Effect-operation argument** → [#754](https://github.com/aallan/vera/issues/754)
- **Generic-instantiated `@Int` field** (`Some(@Nat.0)` into `Option<Int>`) → [#757](https://github.com/aallan/vera/issues/757)
- **Tuple construction / destructure component** → [#758](https://github.com/aallan/vera/issues/758)

## Remaining scope (this issue)

The widening sites with no existing narrowing-dual home, all blocked on the same missing per-component target-type metadata in code generation:

- **Array-literal element** — `let @Array<Int> = [@Nat.0]`. The literal is typed by its element *values* (source) and the `let` is a pointer-pair copy, so the element coercion has no guard site (`_infer_array_element_type` returns the source type). Currently disclosed as an unguarded **E531** warning. (The `@Int → @Nat` array-element narrowing is not separately tracked, so it belongs here too.)
- **Closure argument / return / capture** — `apply_fn` is excluded from the verifier's call-arg path and the `call_indirect` codegen has no guard machinery; closure bodies are opaque to the verifier. Needs closure formal-type recovery (same blocker as the narrowing direction).
- **Heterogeneous `if`/`match` with a genuine `@Int`-*slot* alternative** — `if c then { @Nat.0 } else { @Int.x }`. The literal case is fixed (#821); a genuine `@Int` slot makes the join genuinely `@Int`, so the boundary guard cannot fire (it would false-trap a negative `@Int.x`) and per-arm guarding needs the if/match join type — which the checker's sparse side-tables do not record and code generation lacks entirely. The `@Nat` arm widens silently on an unverified compile.

The governing rule (spec §2.2.1, §11.2.1): a widen guard traps on a negative i64, so it may fire **only** when the source is provably `@Nat` — never on a genuine `@Int`, which can be legitimately negative.

## Notes

Each item is test-first (a `vera run -- 18446744073709551615` repro that returns `-1` is the RED case) and should reuse the existing `#813` machinery: `_emit_int_widen_guard` / `_result_is_nat` / `_check_int_widening_obligation(…, guarded=)` / `E530` / `E531`, plus the verifier↔codegen differential (`tests/test_int_widening_differential.py`) extended to the newly-guarded sites. The unifying fix is to thread per-component target-type metadata into code generation, which would unblock both this issue and the narrowing duals (#754, #757, #758) together.

### PR DESCRIPTION
**Pull request by aallan**
## Summary

Fixes #820 — the last of the seven original v0.1.4 burndown bugs. The `@Nat` → `@Int` widening residuals that carried **no obligation, no guard, and no E531 disclosure** widened silently: a closure argument/return/capture, and a heterogeneous `if`/`match` whose alternative arm is a genuine `@Int` slot, all bit-reinterpreted values above `2^63 - 1` to negative `@Int` on `verify`-clean programs (pre-fix evidence: eight sites, each verify-clean or E531-disclosed, each returning `-1` for the `u64.MAX` sentinel). The tuple/array-element component coercions were E531-disclosed but runtime-unguarded.

## The enabler

The checker already recorded `expr_target_types` (the expected type each expression was checked against) but passed it only to the verifier. This PR threads it into **codegen** — `compile()` → `CodeGenerator` → `WasmContext` — giving the WASM layer the same per-component target-type knowledge the verifier has, the dual of the verifier's `_target_type_of`. Closure formal types are recovered identically on both sides through the shared alias resolver.

## The sites — every one obligated AND guarded, differential-enforced

Array elements, tuple construction and destructure, heterogeneous `if`/`match` arms (per-arm, under path/arm conditions), and closure argument/return/capture: the verifier emits the `nat_to_int_coerce` obligation and codegen emits the boundary trap. Post-fix, every site verifies with an honest Tier-3 runtime check, traps at `u64.MAX`, and passes normal values untouched. The extended `test_int_widening_differential.py` (24 cases) pins the two site sets against each other; 13 independent mutations (per-guard, per-obligation, and the enabler threading itself) each turn exactly their targeted tests RED.

## Corpus fallout

Examples unchanged (pin holds at 284/96/380, empirically). Six conformance programs gain honest Tier-3 obligations — all heterogeneous-match shapes (`Ok(@String) -> string_length(...)` with an `Err(_) -> 0 - 1` arm, and `Some(@Nat) -> nat_to_int(...)` siblings) — none go verify-RED, all still execute correctly.

## Honest scope notes

- The generic-instantiated `@Int` field (`Some(@Nat.0)` into `Option<Int>`) **stays E531-disclosed** — it is #757's dual (erased to i64, needs per-field monomorphization metadata); the Tuple guard is explicitly gated so this site keeps its disclosure rather than gaining a partial guard.
- The enabler unblocks the #754/#757 limitations (recorded in their rows; they stay open limitations). Contrary to the plan's original grouping, **#765 is not unblocked** — it is a pattern-extraction-depth issue orthogonal to target-type metadata.
- A nested closure's widening return is codegen-guarded but not verifier-obligated (the `AnonFn` handler is deliberately shallow to avoid false Tier-1s) — sound over-guarding, filed as [#985](https://github.com/aallan/vera/issues/985) (reporting-completeness, #779 family).
- The closure-signature metadata is the enabler for #984's closure return-narrowing fix (queued separately).

## Docs lockstep

#820 rows deleted from KNOWN_ISSUES.md and SKILL.md; the #754/#757 row rewritten to its unblocked-but-open state; E531 disclosure text in spec/02, spec/11, and SKILL narrowed to the sole surviving generic-field site; CHANGELOG `[Unreleased]` bullet; counts reconciled including the gate-blind sites; site assets regenerated.

Closes #820

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved `@Nat` → `@Int` widening runtime guarding at additional coercion sites, including array elements, tuple construction/destructuring, heterogeneous `if`/`match` arms, and closure argument/return boundaries—avoiding incorrect traps on genuine `@Int` paths.
  * Added per-component target-type handling and tightened verifier↔codegen agreement; imported-function widening guards follow the documented limitation.
* **Documentation**
  * Refined the `Int`/`Nat` guidance, `@Nat`→`@Int` spec, and known issues/roadmap/testing notes for the expanded guarding coverage and remaining gaps.
* **Tests**
  * Expanded regression, differential, and CLI/tester end-to-end coverage (including tail-call/heterogeneous desynchronisation scenarios).

## Review round (6be7f2a)

The pr-review agents + adversarial panel confirmed five defects in the opened version, all fixed RED-first and mutation-validated in `6be7f2a` (full record in the review-record comment below): the per-arm widen guard was dead code under a tail-call arm (verify said Tier-3-guarded, run returned `-1`); the per-arm gate was target-blind and **false-trapped** a legal `@Nat` in a `@Nat`-returning hetero join (regression vs base); `vera test` compiled without the artifact tables; a user `data Tuple` fooled the builtin-carrier name gate; and an engineered cross-file span collision could emit a spurious guard on an imported body (imported decls are now excluded from span-keyed recovery). Coverage gaps closed alongside: CLI-threading pins, both untested arm positions, and `2^63 − 1` boundary controls at all six new sites.

Two residuals are filed rather than hidden: #985 (nested-closure obligation under-reporting — sound over-guarding) and #987 (the array/tuple guards are not yet *emitted* for imported module bodies; disclosed in KNOWN_ISSUES/SKILL/spec/CHANGELOG, queued in the burndown).

### CODE REVIEW

**Comment by aallan** [comment]
## Review round — 3 pr-review agents + 4-lens adversarial panel

Per the burndown's standing process, this PR ran the full gauntlet: code-reviewer, pr-test-analyzer, and comment-analyzer agents, plus an adversarial panel (boundary-value, entry-point, corpus-differential, and interaction lenses; every finding independently verified by two skeptics — 4 confirmed, 0 rejected, 51 clean checks). All confirmed findings are fixed in `6be7f2a`.

### Soundness fixes (all RED-first, mutation-validated)

1. **Per-arm widen guard was dead code under a tail-call arm** (code-reviewer + panel, independently). A heterogeneous `@Int`-join arm whose `@Nat` value is a call in tail position lowered to `return_call`; the guard appended after the terminator never ran — `vera verify` reported the site Tier-3 runtime-guarded while `vera run` returned `-1` silently. Fixed by collecting the call ids under guarded arms (`_collect_hetero_widen_arm_calls`, mirroring #983's narrowing-leaf machinery) and subtracting them from the tail-call set: the guarded arm's call lowers to a plain `call` with a live guard, and the genuine `@Int` arm's recursive `return_call` keeps TCO (pinned at 100k depth). `tests/test_hetero_widen_tailcall.py`.
2. **The per-arm gate was target-blind — a false-trap regression** (panel, interaction lens). The gate never required the join to *target* `@Int` (the verifier's `_is_hetero_int_widen_join` does), so a hetero join in a `@Nat`-returning function widen-guarded its legal `@Nat` arm: a verify-clean Tier-1 program trapped on values above `2^63 - 1` where base returned them. A shared `_is_hetero_int_widen_join` now gates both emitters and the tail-call collector.
3. **`vera test` compiled and verified without the artifact tables** (code-reviewer + panel). The new guards vanished from tester-compiled WASM while the verifier claimed Tier-3 (latent in practice — the tester bounds inputs to ±2^53 — but a broken invariant). `cmd_test` now uses `typecheck_with_artifacts` and threads both tables through the Tester. `tests/test_tester_artifacts.py`.
4. **A user `data Tuple<A, B>` fooled the builtin-carrier name gate** (code-reviewer). Codegen guarded a construction the verifier never obligated (opposite-direction desync; low reachability). The gate now keys on the variadic carrier's empty layout, not the constructor name.
5. **Span-collision false guard for imported bodies** (panel probe, engineered repro). The target table is span-keyed with no file identity; a line-for-line collision made an imported all-`@Nat` function pick up a main-file `Int` target and false-trap. Imported decls are now excluded from span-keyed recovery (`tests/test_xmod_span_collision.py`).

### Residuals — filed, not hidden

- **#987** (new, queued in the burndown): the array/tuple component guards are still not *emitted* for imported module bodies — the single-module table can't serve them, so a library's own Tier-3 promise is dropped through the import door. Disclosed in KNOWN_ISSUES, SKILL, spec/11, and the CHANGELOG; the fix is per-module artifact threading with file-aware keying.
- **#985** (filed during this PR): a closure nested inside another closure's body is guarded by codegen but not obligated by the verifier — a reporting-completeness gap, sound over-guarding.

### Test-coverage round (pr-test-analyzer)

CLI-level pins added (the corpus previously threaded the tables by hand, so deleting the `vera run`/`vera compile` threading broke nothing): `vera run` traps at `u64.MAX` with an in-range control, `--wat` carries the guard with a no-widening control, `verify --json` surface pin. Two honest deviations: the `cmd_verify` threading mutation is unkillable — `verify()` self-heals by recomputing the table (#747 fallback) — and a request-driven `cmd_serve` trap test is infeasible without faking (three documented reasons). Also added: middle-match-arm and else-if-arm differential rows (the committed corpus only exercised one arm position per construct) and `2^63 - 1` boundary controls at all six new sites.

### Panel clean checks (highlights of 51)

Boundary matrix exact at all 8 guarded sites (trap threshold precisely the sign bit; `2^63 - 1` passes untouched; genuine negative `@Int` controls never trap); corpus differential over all 187 examples + conformance programs at base vs head shows exactly the six expected Tier-3 conformance deltas and zero other drift (examples aggregate 284/96/380 unchanged, all negative fixtures identical); the disclosed generic-field site (#757 dual) behaves identically at both commits; closure widen/narrow directions not conflated (#984 stays as filed); no double-emission of narrow+widen guards on one value.

Docs-accuracy round: stale guarded/unguarded enumerations in the verifier docstrings and the user-facing E531 rationale updated to post-#820 reality; `#758 widening residual` mis-references corrected to #813; spec/11 closure-recovery attribution fixed; counts reconciled (7,129 tests / 109 files).

File: tests/test_nat_int_widening.py [hunk 1]
```diff
@@ -444,3 +446,130 @@ def test_all_int_if_emits_no_coerce_obligation(self) -> None:
...
+public fn f(@Nat -> @Int)
+  requires(true) ensures(true) effects(pure)
+{ let @NatToInt = fn(@Nat -> @Int) effects(pure) { @Nat.0 }; apply_fn(@NatToInt.0, @Nat.0) }
+""")
+        co = [o for o in result.obligations if o.kind == _KIND]
+        # One for the closure return (the @Nat body) + one for the @Nat argument.
+        assert any(o.status == "tier3" for o in co), [
+            (o.kind, o.status) for o in result.obligations]
+        assert all(o.status == "tier3" for o in co), [
+            (o.kind, o.status) for o in co]
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_nat_int_widening.py]: @coderabbitai Fixed in 6be7f2a — the comment now describes the single return-side obligation (the closure formal is @Nat, so the apply_fn argument is Nat→Nat and does not widen), and the loose any/all is replaced with a len(co) == 1 pin plus a tier3 status assert, mirroring the sibling capture test.

**Comment by aallan** [comment]
@coderabbitai The outside-diff vera/cli.py finding (vera/tester.py compiling without expr_target_types) is fixed in 6be7f2a: cmd_test now runs typecheck_with_artifacts and threads both tables through the Tester into its verify() and codegen_compile() calls, so vera test compiles with the same per-component guards as the compile/run/serve doors. Pinned by tests/test_tester_artifacts.py — the tester-compiled WAT for a tuple-component widening must carry the widen guard (mutation-killing for the threading kwarg).

File: SKILL.md [hunk 2]
```diff
@@ -412,7 +412,7 @@ array, useful when processing diagnostics programmatically.
...
+**`Int` and `Nat` are interchangeable in both directions.**  `@Nat <: @Int` is a formal subtyping rule at the *type* level (use a `@Nat` anywhere `@Int` is expected, no `nat_to_int` call), and `@Int <: @Nat` is permitted by the type checker with a verifier-discharged obligation (`@Int.0 >= 0`).  This means `array_length` (declared `@Int`) flows freely into `@Nat` positions without explicit conversion — the verifier proves non-negativity from context or falls back to a runtime check.  Both directions carry a *value*-level obligation, because `@Nat` is a u64 and `@Int` an i64: narrowing requires `>= 0` (`E503`/`E504`), and widening requires `<= i64.MAX` (`E530`; or an `E531` warning at the generic-instantiated `@Int`-field component site code generation cannot guard) — a `@Nat` above i64.MAX bit-reinterprets to a negative `@Int`.  **Do not** insert `nat_to_int` defensively; it's only needed when the value is genuinely allowed to be negative.  See spec §2.2.1 for the formal rule.
```
[the remark this answers is not in the document]
Comment (aallan) [reply SKILL.md]: @coderabbitai Clarified in 45e5afd — the paragraph now states the one-to-one guard/obligation correspondence explicitly and names its single exception: a closure nested inside another closure's body is runtime-guarded but not yet obligated (#985), so a guard can fire there that vera verify did not report. The 6be7f2a disclosures (KNOWN_ISSUES #985 row + CHANGELOG caveat) carry the mechanism detail.

File: spec/11-compilation.md [hunk 3]
```diff
@@ -71,7 +71,7 @@ Arithmetic **overflow** of `@Int`/`@Nat` addition, subtraction, and multiplicati
...
...heir `@Int` target; the closure positions recover it from the closure's own function type).  The table-recovered component guards currently apply to the **top-level module only**: an imported function's array-element / tuple-component widening is not yet runtime-guarded through the import door, though the library's own `vera verify` reports the site Tier-3 — tracked as [#987](https://github.com/aallan/vera/issues/987) (the table is span-keyed and single-module; imported bodies are excluded from span-keyed recovery rather than risking a colliding span's wrong target).  The one *component* coercion site code generation still cannot guard is a generic-instantiated `@Int` field (e.g. `Some(@Nat.0)` into `Option<Int>`, erased to i64 with no per-field mono metadata): the widening there is disclosed as an unguarded **E531** warning rather than claiming a runtime check it never emits (tracked with its narrowing dual, [#757](https://github.com/aallan/vera/issues/757)).  Unlike the narrowing guard (a no-op on a valid `@Nat`, so safe to apply to every `@Nat` target), the widen guard fires **only** when the source is provably `@Nat` — never a genuine `@Int`, which may be legitimately negative.
```
[the remark this answers is not in the document]
Comment (aallan) [reply spec/11-compilation.md]: @coderabbitai Fixed in 1f63d7e — the enumeration now reads "closure argument, return, and capture positions (a captured @Nat widening into an @Int closure body shares the body-return guard)", matching the CHANGELOG's return/capture pairing.

File: tests/test_hetero_widen_tailcall.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    result = _compile_with_types(source)
+    exec_result = execute(result, fn_name=fn, args=args)
+    assert exec_result.value is not None
+    return exec_result.value
+
+
+def _assert_traps(source: str, fn: str, args: list[int]) -> None:
+    result = _compile_with_types(source)
+    with pytest.raises((wasmtime.WasmtimeError, wasmtime.Trap, RuntimeError)):
+        execute(result, fn_name=fn, args=args)
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_hetero_widen_tailcall.py:92]: @coderabbitai Fixed in 1f63d7e — _assert_traps now catches the normalized WasmTrapError and pins exc_info.value.kind == "unreachable" (the widen guard is a bare unreachable net), so the assertion proves the guard fired rather than any wasm failure.

File: tests/test_hetero_widen_tailcall.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+class TestFix3UserTupleGate:
+    def test_user_tuple_verifier_emits_no_guarded_coerce(self) -> None:
+        # The verifier does not runtime-guard the user-Tuple construction.
+        assert "tier3" not in _coerce_statuses(_FIX3_USER_TUPLE)
+
+    def test_user_tuple_run_does_not_trap(self) -> None:
+        # BUG at head: codegen emitted a widen guard the verifier never
+        # obligated, so run(u64.MAX) trapped.  After the fix the user Tuple's
+        # generic field stays unguarded and the value round-trips.
+        assert _run(_FIX3_USER_TUPLE, "f", [U64_MAX]) is not None
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_hetero_widen_tailcall.py]: @coderabbitai Fixed in 1f63d7e — the test now routes through _assert_no_trap(_FIX3_USER_TUPLE, "f", [U64_MAX], U64_MAX), i.e. the masked round-trip value is compared bit-exactly; a silent truncation would now fail.

File: tests/test_tester_artifacts.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+class TestTesterThreadsWideningArtifacts:
+    def test_tester_compiled_wat_has_tuple_widen_guard(self, monkeypatch) -> None:
+        # BUG at head: cmd_test compiled without the target-type table, so the
+        # tuple-component widen guard (`i64.lt_s`) was absent from the tester's
+        # WASM while the verifier claimed the site tier3-guarded.
+        wat = _cmd_test_capturing_wat(_TUPLE_WIDEN, monkeypatch)
+        assert "i64.lt_s" in wat, (
+            "the tester-compiled WASM is missing the @Nat->@Int tuple-component "
+            "widen guard — cmd_test did not thread expr_target_types into codegen"
+        )
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_tester_artifacts.py:89]: @coderabbitai Fixed in 1f63d7e — the assertion is scoped with the same paren-matching _wat_function slicer test_cli.py uses, so it checks i64.lt_s inside $tc specifically rather than anywhere in the module.

File: tests/test_xmod_span_collision.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        )
+
+    def test_imported_body_not_false_guarded_by_collision(self, tmp_path) -> None:
+        # BUG at head: the imported all-@Nat `lw` picked up main's
+        # `Tuple<Int, Int>` target at the colliding span and emitted a spurious
+        # widen guard, so `callit` (which calls `lw(u64.MAX)`) trapped.  After
+        # the fix the imported body ignores the main span table — no trap.
+        result = _compile_main(tmp_path)
+        exec_result = execute(result, fn_name="callit", args=[])
+        assert exec_result.value is not None
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_xmod_span_collision.py:113]: @coderabbitai Fixed in 1f63d7e — the collision test now asserts exec_result.value & ((1 << 64) - 1) == U64_MAX (lw is the @Nat identity), so a silent corruption can no longer pass as "no trap".

File: tests/test_xmod_span_collision.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+    def test_same_file_top_level_guard_unaffected(self, tmp_path) -> None:
+        # Control: the SAME-file `mw` genuinely targets `Tuple<Int, Int>`, so its
+        # @Nat component widen guard must remain — `mw(u64.MAX)` still traps.
+        result = _compile_main(tmp_path)
+        with pytest.raises((wasmtime.WasmtimeError, wasmtime.Trap, RuntimeError)):
+            execute(result, fn_name="mw", args=[U64_MAX])
```
[the remark this answers is not in the document]
Comment (aallan) [reply tests/test_xmod_span_collision.py:121]: @coderabbitai Fixed in 1f63d7e — the control now catches WasmTrapError and pins kind == "unreachable", same tightening as the sibling helper in test_hetero_widen_tailcall.py.

**Comment by aallan** [comment]
@coderabbitai The outside-diff test_nat_int_widening.py finding (pin the obligation counts in the three guarded tuple/array tests) is addressed in 1f63d7e, with one measured correction to the suggestion: test_array_element_widening_tier3_guarded and test_tuple_construction_component_tier3_guarded pin len(co) == 1 as suggested, but test_tuple_destructure_literal_tier3_guarded pins len(co) == 2 — the fixture destructures Tuple(@Nat.0, @Nat.0) into two @Int slots, and each @Nat component carries its own obligation (verified empirically before pinning; the suggested len == 1 would have been wrong there). The counts also answer the construction+destructure double-count question: construction and destructure are distinct sites with distinct obligations, and each test's count is now pinned exactly.
